x-pack/plugins/security_solution/public/common/components/toolbar/bulk_actions/alert_bulk_tags.tsx

```diff
@@ -0,0 +1,170 @@
+import { useUiSetting$ } from '../../../lib/kibana';
+import { useSetAlertTags } from './use_set_alert_tags';
+import { useAppToasts } from '../../../hooks/use_app_toasts';
+import * as i18n from './translations';
+import { createInitialTagsState } from './helpers';
+
+interface BulkAlertTagsPanelComponentProps {
+  alertItems: TimelineItem[];
+  refetchQuery?: () => void;
+  setIsLoading: (isLoading: boolean) => void;
+  refresh?: () => void;
+  clearSelection?: () => void;
+  closePopoverMenu: () => void;
+}
+const BulkAlertTagsPanelComponent: React.FC<BulkAlertTagsPanelComponentProps> = ({
+  alertItems,
+  refresh,
+  refetchQuery,
+  setIsLoading,
+  clearSelection,
+  closePopoverMenu,
+}) => {
+  const [defaultAlertTagOptions] = useUiSetting$<string[]>(DEFAULT_ALERT_TAGS_KEY);
+  const { addSuccess, addError, addWarning } = useAppToasts();
+
+  const { setAlertTags } = useSetAlertTags();
+  const existingTags = useMemo(
+    () =>
+      alertItems.map(
+        (item) => item.data.find((data) => data.field === ALERT_WORKFLOW_TAGS)?.value ?? []
+      ),
+    [alertItems]
+  );
+  const initalTagsState = useMemo(
+    () => createInitialTagsState(existingTags, defaultAlertTagOptions),
+    [existingTags, defaultAlertTagOptions]
+  );
+
+  const tagsToAdd: Set<string> = useMemo(() => new Set(), []);
+  const tagsToRemove: Set<string> = useMemo(() => new Set(), []);
+
+  const onUpdateSuccess = useCallback(
+    (updated: number, conflicts: number) => {
+      if (conflicts > 0) {
+        addWarning({
+          title: i18n.UPDATE_ALERT_TAGS_FAILED(conflicts),
+          text: i18n.UPDATE_ALERT_TAGS_FAILED_DETAILED(updated, conflicts),
+        });
+      } else {
+        addSuccess(i18n.UPDATE_ALERT_TAGS_SUCCESS_TOAST(updated));
+      }
+    },
+    [addSuccess, addWarning]
+  );
+
+  const onUpdateFailure = useCallback(
+    (error: Error) => {
+      addError(error.message, { title: i18n.UPDATE_ALERT_TAGS_FAILURE });
+    },
+    [addError]
+  );
+
+  const [selectableAlertTags, setSelectableAlertTags] =
+    useState<EuiSelectableOption[]>(initalTagsState);
+
+  const onTagsUpdate = useCallback(async () => {
+    closePopoverMenu();
+    const ids = alertItems.map((item) => item._id);
+    const query: Record<string, unknown> = getUpdateAlertsQuery(ids).query;
+    const tagsToAddArray = Array.from(tagsToAdd);
+    const tagsToRemoveArray = Array.from(tagsToRemove);
+    try {
+      setIsLoading(true);
+
+      const response = await setAlertTags({
+        tags: { tags_to_add: tagsToAddArray, tags_to_remove: tagsToRemoveArray },
+        query,
+      });
+
+      setIsLoading(false);
+      if (refetchQuery) refetchQuery();
+      if (refresh) refresh();
+      if (clearSelection) clearSelection();
+
+      if (response.version_conflicts && ids.length === 1) {
+        throw new Error(i18n.BULK_ACTION_FAILED_SINGLE_ALERT);
+      }
+
+      onUpdateSuccess(response.updated ?? 0, response.version_conflicts ?? 0);
```

Comment:
We are missing checks around the async updates to see if the component is still mounted before making state updates. This will cause unintended side effects and memory leaks.

There is a hook you can use called [`useFetch`](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/public/common/hooks/use_fetch/use_fetch.ts#L114). It is a generic hook that simplifies the async request queries implementation. It provides APM monitoring, abort control, error handling and refetching. 

If you choose not to use this hook, please at least follow the abort/ignore pattern it provides. Always check the `ignore` value before making state updates

Reply:
Yes, I think `useFetch` would be helpful here. It would look something like this:

```
  const {
    fetch,
    isLoading,
    data: response,
    error,
  } = useFetch(REQUEST_NAMES.SET_ALERTS_TAGS, setAlertTags);
```
Then inside the `onTagsUpdate` callback, we'd just need to replace the async logic by:
`fetch({ tags: { tags_to_add: tagsToAddArray, tags_to_remove: tagsToRemoveArray }, query });`

The toasts logic will need to be moved inside a `useEffect`